### ISSUE TITLE
Buffering

### ISSUE THREAD

**Issue by atifaziz** [morelinq/MoreLINQ#87]
The use of buffering on `Repeat()` and `Cartesian()` is currently commented as a TODO. The attached patch:
1. Implements a buffering class that buffers enumerables in memory.
2. Changes `Repeat()` and `Cartesian` to use this buffering.
3. Further changes `Repeat()` to skip this if the count is 0 or 1.
4. Defines an `IBufferedEnumerable<T>` interfaces, documented as adding a buffering guarantee to `IEnumerable<T>`.
5. Defines two overloads of a method to create such a buffered instance. One of which insists that buffering happen even if the source implements `ICollection<T>`. (Otherwise it is assumed that the source already serves as a buffer adequately).

In short: Deals with the two TODO items, and brings the advantage of the code needed to do so, to the user.

---

Originally reported on Google Code with ID 87

Reported by `jonhanna76.JH` on 2013-12-12 14:46:28

---
- _Attachment: [Buffering.patch](https://storage.googleapis.com/google-code-attachments/morelinq/issue-87/comment-0/Buffering.patch)_

**Comment by fsateler** [comment]
This is the same as #100
